### ISSUE TITLE
Doppler shift and wavelength axis

### ISSUE THREAD

**Issue by bgrierson21** [D3DEnergetic/FIDASIM#48]
This is an issue regarding the definition of the wavelength axis and where on the lambda grid photons are accumulated.  In the IDL version of FIDAsim the photons were accumulated on wavelength grid "boundaries", not grid "centers".  In order to properly interpret the doppler shift from the halo spectrum the doppler shift had to be calculated with the wavelength axis shifted by "delta lambda" divided by two, where "delta lambda" is the wavelength grid spacing.  I have run cases with flat ne, Te, Ti, Zeff profiles and zero rotation.  Three runs were executed with wavelength nlambda=768 (default), as well as double and half of this value.  If the halo spectrum is turned into a doppler velocity using the output "lambda", then the velocity is finite and scales with the nlambda, which is incorrect.  If the halo spectrum is turned into a velocity using lambda+dlambda/2 then the velocity comes out to approximately zero for all three cases.  Therefore, the proper wavelength axis for plotting and fitting of FIDASIM output is to define lambda = result.spectra.lambda + (lambda[1]-lambda[0])/2.0.  Comments and/or confirmation that this is the proper interpretation is welcome.

**Comment by lstagner** [comment]
Im curious how you are calculating this. When I go from wavelength to velocity using the standard definition of doppler shift I get that there is little difference.
![screenshot from 2015-03-11 11 24 15](https://cloud.githubusercontent.com/assets/4305891/6603779/6b84892e-c7e1-11e4-8140-34ccd1e4c337.png)

**Comment by bgrierson21** [comment]
There may be little difference to the eye but you just plotted a doppler shift that is a not-so-insignifincat fraction of the speed of light, corresponding to 75 km/s line-of sight velocity.  That's huge.

**Comment by lstagner** [comment]
Alight here is the section of the code that bins the calculated wavelength
![screenshot from 2015-03-11 12 27 54](https://cloud.githubusercontent.com/assets/4305891/6605048/2d2310fc-c7ea-11e4-8b29-86a145040733.png)

How its calculating the wavelength bin is not correct. If the bins are defined by their centers, it should be
```fortran
bin=floor(((wavel(i)-(spec%lambdamin-spec%dlambda/2.0 )/spec%dlambda)) + 1
```

**Comment by bgrierson21** [comment]
The critical question for the developers and users is whether to change it or leave it.  Either way this needs to be communicated because the Doppler shifts from what I've seen are significant for Doppler shift spectroscopy, but not for overplotting, say, FIDA radiance on top of experimental data.  This issue needs to be elevated.

**Comment by lstagner** [comment]
Seems ```spec%lambdamin``` already include the bin width since the lambda arrays are calculated with

```fortran
do i=1,spec%nlambda 
       lambda_arr(i)=(i-0.5)*spec%dlambda*0.1d0 +spec%lambdamin*0.1d0
enddo 
```
so the code is correct. Regardless, a user should not have to know internal details about fidasim to do a simple calculation. We should change it so that obvious thing is the right thing. I.e. they shouldn't have to add ```dwav/2``` to lambda when converting from wavelength to velocity.

**Comment by bgrierson21** [comment]
Then someone with the proper knowledge of the fortran should either appeal, or change it and double check that it propagates correctly.  I'd rather not have a conditional on the version of FIDASIM that I need to check in post-processing.

**Comment by lstagner** [comment]
We could have it such that FIDASIM calculates the velocities for the user (relative to 656.28 nm)

**Comment by bgrierson21** [comment]
I disagree.  The user should deal with this and the possibly of non-gaussian looking spectra in the presence of steep gradients.  Also, I believe that rest wavelength in fidasim.f90 is 6561.0 angstroms defined as real(double),parameter:: lambda0=6561.0d0 !! D-alpha[A]

**Comment by lstagner** [comment]
Ok I think I found the problem. You shouldn't have to add ```dwav/2``` to the wavelengths. The issue is that FIDASIM is incorrectly binning the wavelengths. The problem is the the fortran ```int``` acts like a ```floor``` instead of like ```round``` like I previously assumed. The result is that some wavelengths get binned to the next highest wavelength bin. This demonstrates the issue.

![screenshot from 2015-03-11 14 37 56](https://cloud.githubusercontent.com/assets/4305891/6607498/9892dc7a-c7fc-11e4-91af-73414e441338.png)

I will fix this in my interpolation branch.

**Comment by lstagner** [comment]
![image](https://cloud.githubusercontent.com/assets/4305891/6608699/8306fcd8-c806-11e4-8bd9-e76b1bec93b7.png)
In regards to wavelength outside the grid right now we tack them on to closest wavelength. Should we just discard them instead?

**Comment by geigerb** [comment]
Luke and Brian, 
I agree that there was a bug! 
the code should be
bin=floor((wavel(i)-diag%lambdamin)/diag%dlambda)+1

**Comment by bgrierson21** [comment]
I'm using FIDASIM for some studies right now that I'll have to re-do once this correction is made.  When are we going to have a new version of the fortran that I can pull?  I think that photons that land outside of the spectral range should be discarded.

**Comment by lstagner** [comment]
Brian,
The fix is easy to implement. If i were you, I would make the necessary changes in a new local branch, commit those changes, and use that branch for your analysis. That is what I do when I need to make custom changes to the source code for some analysis, for instance my [ls/srs_bench](https://github.com/D3DEnergetic/FIDASIM/tree/ls/srs_bench) branch.

**Comment by bgrierson21** [comment]
@geigerb is that the exact syntax to replace the line that begins with "bin="?

**Comment by lstagner** [comment]
@bgrierson21 Here just copy [this](https://github.com/D3DEnergetic/FIDASIM/commit/89e305ae38c9b23c93cb3417158100589ac64529)

**Comment by lstagner** [comment]
@geigerb 's version of FIDASIM uses different names for structures. What he calls ```diag``` the official version calls ```spec```. So you should use ```spec```.

**Comment by geigerb** [comment]
... sorry for the confusion! I used the name 'diag' insted of 'spec' without any good reason!

**Comment by lstagner** [comment]
Fixed in release-0.3 branch
